Add documentation for Arduino MKR WAN 1310 (#1343)
TheThingsIndustries/lorawan-stack-docs: doc/content/devices/models/mkr1310/temp.md

---
title: "Sending Temperature with MKR WAN 1310"
description: ""
weight:
---

In this section you are going to try an example project where you will connect a Temperature and Humidity sensor to the MKR WAN 1310 to send the data over LoRaWAN and then decode it on {{% tts %}}.

<!--more-->

{{< figure src="../fritzing.png" alt="Wiring mockup of the DHT11 connected to the arduino" class="float plain" width="90%">}}

> ## Circuit

Reviewer: ```suggestion
## Building the Circuit
```

For this project you are going to be needing:
- **DHT11**
- Some **Jumper Wires**
- **Breadboard**

Once you have the components, wire it up like so:
- **Arduino 5V** -> **Plus DHT11**
- **Arduino GND** -> **Minus DHT11**
- **Arduino Pin 6** -> **Signal DHT11**

> Illustration on the right displays the wiring as well.

Reviewer: remove

{{< note "Double check your own sensor's pinout, it can vary between modules. If you are using a bare DHT11, you might need to use a resistor." />}}

> ## Creating the code

Reviewer: ```suggestion
## Programming the MKR 1310
```

Reviewer: I think this whole section is overkill. It would be better to just paste the complete code and eventually add comments where needed

Author: Alright fair. I just thought it'd be good to explain to the users what the code is doing exactly.

> Before we continue, you need to make sure you have the **DHT sensor library by Adafruit** installed: 

Reviewer: ```suggestion
Before we continue, you need to make sure you have the **DHT sensor library by Adafruit** installed in the Arduino IDE:
```

  modem.write(payload, sizeof(payload));
  modem.endPacket(true);
  delay(100000);
}
```

### Entire code

```cpp
#include <MKRWAN.h>
#include <DHT.h>

LoRaModem modem(Serial1);

String appEui = "";
String appKey = "";

DHT dht(6, DHT11);

void setup() {
  // put your setup code here, to run once:
  Serial.begin(115200);
  while (!Serial);
  // change this to your regional band (eg. US915, AS923, ...)
  if (!modem.begin(EU868)) {
    Serial.println("Failed to start module");
    while (1) {}
  };
  
  dht.begin();

  int connected = modem.joinOTAA(appEui, appKey);
  if (!connected) {
    Serial.println("Something went wrong; are you indoor? Move near a window and retry");
    while (1) {}
  }
}

void loop() {
  int t = dht.readTemperature() * 100;
  int h = dht.readHumidity() * 100;

  byte payload[4];
  payload[0] = highByte(t);
  payload[1] = lowByte(t);
  payload[2] = highByte(h);
  payload[3] = lowByte(h);

  Serial.println("Temperature: ");
  Serial.println(dht.readTemperature());
  Serial.println("Humidity: ");
  Serial.println(dht.readHumidity());

  modem.beginPacket();
  modem.write(payload, sizeof(payload));
  modem.endPacket(true);
  delay(100000);
}
```

> ## Uploading the code

Reviewer: No need for this to be a separate section from programming the device

> Now to upload the code to the MKR WAN 1310. First plug the Arduino into your computer. In the IDE, click on **Select Board** at the top, and select your **Arduino MKR WAN 1310**.

Reviewer: ```suggestion
Now the code needs to be uploaded to the MKR WAN 1310. First plug the Arduino into your computer. In the IDE, click on **Select Board** at the top, and select your **Arduino MKR WAN 1310**.
```

{{< figure src="../select-board.png" alt="Selecting Board" >}}

Once that is done you can upload the file to the board (Using the **->** button). Once the code is uploaded successfully, open the Serial Monitor. You can open it by clicking the **icon in the top-right corner**.

{{< figure src="../serial-monitor.png" alt="Serial Monitor button highlighted" >}}

> Here you can check if the temperature and humidity values are showing up correctly.

Reviewer: ```suggestion
Temperature and humidity values should be listed in the Serial Monitor.
```

{{< figure src="../serial-monitor-temp.png" alt="Serial Monitor" width="55%" >}}

> ## {{% tts %}}

Reviewer: Please don't use this as a title, as it literally doesn't explain anything. It would be weird when doing docs search too. You can put something like "Decode temperature and humidity data" or something like that

Now that the device is working go back to {{% tts %}} and set up the payload formatter. A payload formatter will turn the hexadecimal string received into readable text. To do this, in your application on {{% tts %}}, click on **Payload formatters**.

{{< figure src="../stack-pf.png" alt="payload formatter highlighted" width="70%" >}}

> Make sure that **Custom Javascript formatter** is selected here from the dropdown.

Reviewer: ```suggestion
Select **Custom Javascript formatter** from the dropdown.
```

{{< figure src="../custom-js.png" alt="custom javascript selected in dropdown" width="70%"   >}}

> Now in the **Formatter code** field, replace it with this formatter:

Reviewer: ```suggestion
Now paste the following code in the **Formatter code** field:
```

```js
function decodeUplink(input) {
  // Read the temperature and humidity from the payload
  var t = (input.bytes[0] << 8 | input.bytes[1]) / 100;
  var h = (input.bytes[2] << 8 | input.bytes[3]) / 100; 
  
  return {
    data: {
      temperature: t,
      humidity: h
    }
  };
}
```

> Now if you go back to the **Live Data** tab you'll see the temperature and humidity show up!

Reviewer: ```suggestion
If you go back to the **Live Data** tab you'll see the readable temperature and humidity values showing up!
```

{{< figure src="../live-data.png" alt="Live data tab">}}

> This concludes the example project. Now go create your own project!

Reviewer: ```suggestion
This concludes our example project. Now you can proceed with creating your own project! Make sure to check our integrations that can help you visualize your data, set up monitoring and alerting, etc.
```

Reviewer: Link integrations section also

## Troubleshoot

**My DHT sensor does not display any data**
- Make sure that you have wired it up properly.

> - If you are using a bare DHT11, make sure that it has a pull up resistor between Vcc and Signal. Consult google for more info.

Reviewer: Again, too little info for a troubleshoot section. You can instead add a sentence like "If Arduino IDE Serial Monitor doesn't show DHT sensor data, make sure to double-check wiring." to the part where you explain these values showing up in Serial Monitor. And the second point about pull up resistor - it would be good if you can find some example wiring guide and link it in the beginning of this article together with that note about using the resistor. "Consult google" is a bit pretentious - of course everyone trying to do this and having issues will consult Google, and it doesn't add any real value to the article so pls remove

{{< figure src="../fritzing.png" alt="Wiring mockup of the DHT11 connected to the arduino" class="float plain" width="90%">}}

## Building the Circuit

For this project you are going to be needing:
- DHT11 sensor
- Some jumper wires
- Breadboard

Once you have the components, wire it up like shown on the image on the right:
- **Arduino 5V** -> **Plus DHT11**
- **Arduino GND** -> **Minus DHT11**
- **Arduino Pin 6** -> **Signal DHT11**

{{< note "Double check your own sensor's pinout, it can vary between modules. If you are using a bare DHT11, you might need to use a resistor." />}}

## Programming the MKR 1310

Before we continue, you need to make sure you have the **DHT sensor library by Adafruit** installed in the Arduino IDE:  

{{< figure src="../dht-lib.png" alt="DHT sensor library by Adafruit">}}

Next is the code itself. You can find it below, but you do need to change a few things. 

> First you have to **add the AppEUI and AppKey from {{% tts %}} between the double quotes**. 

Reviewer: I actually meant in-code comments, such as:

`String appEui = "123456789ABCDEF";  #Fill in with your AppEUI`

Something like that. Same thing for regional band part

Author: I felt like users would not notice that but sure ill just say "check the in-code comments"
